JAP 3단계 리뷰 요청드립니다.

InmookJeong opened the pull request:
안녕하세요. 우아한테크캠프 Pro 참여자 정인묵입니다.
JAP 3단계 리뷰 요청드립니다.

InmookJeong committed:
docs: modify function list

InmookJeong committed:
refactor: remove auditing annotation

- Removed EntityListeners annotation.(Answer, Question, User)

InmookJeong committed:
test: add answer owner check test code

InmookJeong committed:
test: delete Question and Answers

InmookJeong committed:
feat: add columns wrapper class

InmookJeong committed:
test: delete question and answer

jhhj424 commented on README.md:
```diff
@@ -1,45 +1,67 @@
 ### 1. User
-- 사용자 등록(회원가입)
-- 사용자 정보 수정
+- ~~사용자 등록(회원가입)~~
```
```suggestion
- ~~사용자 등록(회원가입)~~
-[ ] 사용자 등록(회원가입)
-[x] 사용자 등록(회원가입)
```

글자에 줄 긋기로 체크해주셨군요~
둘다 좋은 방법인 것 같아요 :)

기능 목록 꼼꼼하게 작성해주셨네요 💯

jhhj424 commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -1,25 +1,25 @@
-    @Lob
-    private String contents;
+    @Embedded
+    private Contents contents;
```
엄청 꼼꼼하게 `모든 원시값을 포장` 을 진행해주셨네요 👍 
미션중이니 꼼꼼하게 해주신 점 좋은 것 같습니다.

실제로 사용하실때는 특정 기준을 정하고, 해당 기준을 만족하면 객체로 포장하는 방법도 좋은 것 같습니다 :)
ex) 특정 값에 validation 이 필요한경우 객체로 포장해서 해당 객체는 무조껀 Valid 한 객체로 사용하기.

jhhj424 commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -49,11 +49,12 @@ public Answer(Long id, User writer, Question question, String contents) {
 
     public boolean isOwner(User writer) {
-        return this.writer.equals(writer);
+        return this.writer.getUserId().equals(writer.getUserId());
```
```suggestion
        return this.writer.isSameUser(writer);
```

이러한 방식으로 객체에게 물음을 던져보는식으로 생각&연습해보시면 좋을 것 같습니다.

전체적으로 특정 기능이나 역할을 담당할 수 있는 존중받는 객체로 생각하고 책임을 부여하는게 좋을 것 같습니다.

jhhj424 commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -1,25 +1,25 @@
-@EntityListeners(AuditingEntityListener.class)
 @Entity
 @Table(name = "answer")
 public class Answer extends BaseEntity {
```
모든 엔티티 공통 사항으로 `setter` 가 아닌 특정 기능&역할을 드러낼 수 있는 메서드 네이밍을 사용하시면 좋을 것 같네요 :)

그리고 전체적으로 확인해주시고 사용하지 않는 변수나 메서드는 삭제 해주시는게 좋을 것 같습니다~

jhhj424 commented on src/main/java/qna/domain/Answers.java:
```diff
@@ -0,0 +1,81 @@
+ * @since 2021.11.26
+ * @author Inmook,Jeong
+ */
+public class Answers {
```
일급컬렉션을 잘 만들어주셨네요,
그런데 실제로 사용은 제대로 하고계신 것 같지 않은데 확인 해보시면 좋을 것 같습니다.

예를 들어 현재 `User`클래스에서 
`private List<Answer> answers = new ArrayList<Answer>();` 인스턴스 변수가 있는데 `Answers` 일급 컬렉션을 사용하도록 바꿀 수 있을 것 같아요~

참고: https://wbluke.tistory.com/23,   https://jojoldu.tistory.com/559

jhhj424 commented on src/main/java/qna/domain/DeletedHistories.java:
```diff
@@ -0,0 +1,38 @@
+public class DeletedHistories {
+
+	@Autowired
+	DeleteHistoryRepository deleteHistoryRepository;
```
`DeletedHistories`로 선언해주신게 일급컬렉션 도메인 객체인건가요?
도메인 객체가 Repository 를 주입받고있는 구조가 조금 익숙하지 않네요, 다른 방법을 사용해보시면 어떨까요?

그게아니라 혹시 `DeletedHistories` 객체가 `DDD(도메인 주도 설계)`의 `도메인 서비스`의 개념으로 사용하신거라면, 코멘트 남겨주세요 :)

근데 일단 해당 클래스에서 제가 보기에는 주입받고 사용자체를 안하고 있는 것 같네요.
확인이 필요해보입니다~

jhhj424 commented on src/main/java/qna/domain/User.java:
```diff
@@ -78,16 +80,16 @@ public void updatePassword(User loginUser, String newPassword) {
 
     public boolean matchPassword(String targetPassword) {
-        return this.password.equals(targetPassword);
+        return this.password.equals(new Password(targetPassword));
```
이런것도 Password 객체 자체에 targetPassword 와 일치하는지 물음을 던지도록 메서드를 구성해보는게 좋을 것 같아요~

jhhj424 commented on src/main/java/qna/domain/field/Contents.java:
```diff
@@ -0,0 +1,25 @@
+import javax.persistence.Lob;
+
+@Embeddable
+public class Contents {
```
포장해주신 값객체들도 최대한 `setter` 없이 불변으로 사용해보시는걸 추천드립니다.

jhhj424 commented on src/main/java/qna/service/QnaService.java:
```diff
@@ -35,24 +31,46 @@ public Question findQuestionById(Long id) {
+     * @return
+     */
+    public DeletedHistories deleteQuestionWithAnswer(User user, Question question) throws CannotDeleteException {
+        if(!question.getWriter().getId().equals(user.getId())) {
```
참고: https://mangkyu.tistory.com/147

위 글은 디미터의 법칙에 대한 글인데요, 조금 참고해보시면 도움이 되실 것 같습니다.

지금도 잘 작성해주시고 계시지만 조금 더 객체지향적으로 코드를 작성해보는데 도움이 되실 것 같습니다 :)

jhhj424 commented on src/test/java/qna/domain/AnswersTest.java:
```diff
@@ -0,0 +1,58 @@
+import static org.assertj.core.api.Assertions.assertThat;
+
+@DataJpaTest
+public class AnswersTest {
```
해당 테스트 네이밍을 보고 `Answers` 일급컬렉션 객체의 대한 테스트로 착각할 수 있겠네요!

`Repository` 테스트라면 다른 네이밍을 고려해보시면 좋을 것 같습니다.

---

RepositoryTest 외에 Domain 객체 자체 테스트라면 도메인 로직만 테스트하시면 될 것 같습니다.
이 부분은 제가 처음에 체크를 못했던 것 같네요. 🙏

jhhj424 reviewed (changes requested):
안녕하세요 인묵님~ 🙇🏻‍♂️
JPA 3단계 미션도 잘 작성해주셨네요 :)
조금 더 고민해보면 좋을만한 점 코멘트 남겼으니 확인 부탁드리겠습니다.
이번 단계에서 해당 내용들 완료 하고 다음 미션 진행하시면 수월하실 것 같네요 👍 
마지막 단계이니만큼 조금만 더 화이팅 입니다 👊

InmookJeong committed:
refactor: remove not used method and value

InmookJeong committed:
refactor: remove field class setter

InmookJeong committed:
refactor: override user and password equals method

InmookJeong committed:
refactor: rename AnswersDaoTest class

- Before : test.java.qna.domain.AnswersTest.java
- After : test.java.qna.dao.AnswersDaoTest.java

InmookJeong committed:
refactor: modify first class collection mapping

InmookJeong committed:
refactor: question owner check

InmookJeong committed:
refactor: modify setter method name
